lock-api/src/main/java/com/palantir/lock/watch/ClientLockWatchEventLogImpl.java

```diff
@@ -0,0 +1,220 @@
+    private Optional<IdentifiedVersion> latestVersion = Optional.empty();
+
+    public static ClientLockWatchEventLogImpl create() {
+        return create(ClientLockWatchSnapshotUpdaterImpl.create());
+    }
+
+    @VisibleForTesting
+    static ClientLockWatchEventLogImpl create(ClientLockWatchSnapshotUpdater snapshotUpdater) {
+        return new ClientLockWatchEventLogImpl(snapshotUpdater);
+    }
+
+    private ClientLockWatchEventLogImpl(ClientLockWatchSnapshotUpdater snapshotUpdater) {
+        this.snapshotUpdater = snapshotUpdater;
+    }
+
+    @Override
+    public synchronized Optional<IdentifiedVersion> processUpdate(LockWatchStateUpdate update) {
+        final ProcessingVisitor visitor;
+        if (!latestVersion.isPresent() || !update.logId().equals(latestVersion.get().id())) {
+            visitor = new NewLeaderVisitor();
+        } else {
+            visitor = new ProcessingVisitor();
+        }
+        update.accept(visitor);
+        return latestVersion;
+    }
+
+    @Override
+    public synchronized void removeOldEntries(IdentifiedVersion earliestVersion) {
+        Set<Map.Entry<Long, LockWatchEvent>> eventsToBeRemoved =
+                eventMap.headMap(earliestVersion.version()).entrySet();
+        snapshotUpdater.processEvents(
+                eventsToBeRemoved.stream().map(Map.Entry::getValue).collect(Collectors.toList()));
+        eventsToBeRemoved.clear();
+    }
+
+    /**
+     * @param timestampToVersion mapping from timestamp to identified version from client-side event cache;
+     * @param version            latest version that the client knows about; should be before timestamps in the
+     *                           mapping;
+     * @return lock watch events that occurred from (exclusive) the provided version, up to (inclusive) the latest
+     * version in the timestamp to version map.
+     */
+    @Override
+    public synchronized TransactionsLockWatchEvents getEventsForTransactions(
+            Map<Long, IdentifiedVersion> timestampToVersion,
+            Optional<IdentifiedVersion> version) {
+        Optional<IdentifiedVersion> versionInclusive = version.map(this::createInclusiveVersion);
+        IdentifiedVersion toVersion = Collections.max(timestampToVersion.values(), IdentifiedVersion.comparator());
+        IdentifiedVersion currentVersion = getLatestVersionAndVerify(toVersion);
+
+        if (!versionInclusive.isPresent() || differentLeaderOrTooFarBehind(currentVersion, versionInclusive.get())) {
+            return TransactionsLockWatchEvents.failure(snapshotUpdater.getSnapshot(currentVersion));
+        }
+
+        if (eventMap.isEmpty()) {
+            return TransactionsLockWatchEvents.success(ImmutableList.of(), timestampToVersion);
+        }
+
+        IdentifiedVersion fromVersion = versionInclusive.get();
+
+        return TransactionsLockWatchEvents.success(
+                new ArrayList<>(eventMap.subMap(fromVersion.version(), true, toVersion.version(), true).values()),
```

Comment:
factor out constants 'INCLUSIVE', so the reader doesn't have to understand what 'true' means. Also, for stuff like this, do ImmutableMap.copyOf(

Reply:
(because when it can be immutable for free, it should be :))

advantages: frequently immutables style classes will copy to ensure immutability, and ImmutableList.copyOf() is a no-op if you're copying an immutable list.
Also, immutablelist is marginally faster than arraylist because you don't have to do bounds checking. But really immutability is what you want.